README.md
# number-in-words

A Clojure API to convert numbers into English words.

## Usage

Run the project:
- either "lein run"
- or "lein repl" then "(dev)" then "(reset)"
- or "lein uberjar" then "java -jar target/number-in-words.jar"

http://localhost:3000/api-docs/index.html =>Swagger docs

http://localhost:3000/api/34 => returns JSON like {"result":"thirty-four"}

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-10-13
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-10-13
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/number-in-words/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/number-in-words/compare/0.1.0...0.1.1


test/number_in_words/core_test.clj
(ns number-in-words.core-test
  (:require [clojure.test :refer :all]
            [number-in-words.conversion :refer :all :as core]))

(deftest number-conversions
  (testing "Zero to Twenty"
    (is (= "zero" (convert 0)))
    (is (= "one" (convert 1)))
    (is (= "two" (convert 2)))
    (is (= "three" (convert 3)))
    (is (= "four" (convert 4)))
    (is (= "five" (convert 5)))
    (is (= "six" (convert 6)))
    (is (= "seven" (convert 7)))
    (is (= "eight" (convert 8)))
    (is (= "nine" (convert 9)))
    (is (= "ten" (convert 10)))
    (is (= "eleven" (convert 11)))
    (is (= "twelve" (convert 12)))
    (is (= "thirteen" (convert 13)))
    (is (= "fourteen" (convert 14)))
    (is (= "fifteen" (convert 15)))
    (is (= "sixteen" (convert 16)))
    (is (= "seventeen" (convert 17)))
    (is (= "eighteen" (convert 18)))
    (is (= "nineteen" (convert 19)))
    (is (= "twenty" (convert 20))))

  (testing "Round tens under 100"
    (is (= "thirty" (convert 30)))
    (is (= "forty" (convert 40)))
    (is (= "fifty" (convert 50)))
    (is (= "sixty" (convert 60)))
    (is (= "seventy" (convert 70)))
    (is (= "eighty" (convert 80)))
    (is (= "ninety" (convert 90))))

  (testing "Random numbers between 21 and 99"
    (is (= "twenty-five" (convert 25)))
    (is (= "fifty-two" (convert 52)))
    (is (= "seventy-seven" (convert 77)))
    (is (= "eighty-eight" (convert 88)))
    (is (= "ninety-one" (convert 91))))

  (testing "Random round hundreds"
    (is (= "one hundred" (convert 100)))
    (is (= "five hundred" (convert 500)))
    (is (= "nine hundred" (convert 900))))

  (testing "Random numbers between 100 and 999"
    (is (= "one hundred and twenty-three" (convert 123)))
    (is (= "two hundred and fifty-seven" (convert 257)))
    (is (= "eight hundred and eighty-eight" (convert 888)))
    (is (= "nine hundred and ninety-nine" (convert 999))))

  (testing "One thousand"
    (is (= "one thousand" (convert 1000))))

  (testing "Out fo boundaries"
    (is (= out-of-boundary-error-message (convert -1)))
    (is (= out-of-boundary-error-message (convert 1001)))))

(deftest extracting-the-units
  (testing "Extracting the units"
    (let [extract-units #'core/extract-units]
      (is (= 3 (extract-units 3)))
      (is (= 4 (extract-units 14)))
      (is (= 9 (extract-units 99)))
      (is (= 8 (extract-units 188)))
      (is (= 3 (extract-units 1013))))))

(deftest extracting-the-tens
  (testing "Extracting the tens"
    (let [extract-tens #'core/extract-tens]
      (is (= 0 (extract-tens 3)))
      (is (= 10 (extract-tens 13)))
      (is (= 10 (extract-tens 10)))
      (is (= 90 (extract-tens 99)))
      (is (= 0 (extract-tens 101))))))

(deftest extracting-the-hundreds
  (testing "Extracting the hundreds"
    (let [extract-hundreds #'core/extract-hundreds]
      (is (= 0 (extract-hundreds 3)))
      (is (= 0 (extract-hundreds 13)))
      (is (= 100 (extract-hundreds 100)))
      (is (= 100 (extract-hundreds 145)))
      (is (= 200 (extract-hundreds 235)))
      (is (= 200 (extract-hundreds 3235))))))

(deftest round-tens
  (testing "Are they rounds of tens?"
    (let [round-tens? #'core/round-tens?]
      (is (round-tens? 0))
      (is (round-tens? 10))
      (is (round-tens? 50))
      (is (not (round-tens? 11)))
      (is (not (round-tens? -34)))
      (is (not (round-tens? 456))))))

(deftest round-hundreds
  (testing "Are they rounds of hundreds?"
    (let [round-hundreds? #'core/round-hundreds?]
      (is (round-hundreds? 0))
      (is (round-hundreds? 100))
      (is (round-hundreds? 500))
      (is (not (round-hundreds? 110)))
      (is (not (round-hundreds? -340)))
      (is (not (round-hundreds? 1456))))))


project.clj
(defproject number-in-words "0.1.0-SNAPSHOT"
  :description "Translates numbers into grammatically correct English words"
  :url "http://localhost:3000/number-in-words/api"
  :license {:name "Eclipse Public License"
            :url  "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.9.0"]
                 [org.clojure/tools.logging "0.4.1"]
                 [com.stuartsierra/component "0.3.2"]
                 [http-kit "2.3.0"]
                 [metosin/compojure-api "1.1.11"]]
  :main number-in-words.core
  :profiles {:dev     {:source-paths ["dev"]
                       :dependencies [[org.clojure/tools.namespace "0.2.11"]
                                      [reloaded.repl "0.2.4"]]}
             :uberjar {:uberjar-name "number-in-words.jar"
                       :aot          :all}}
  :repl-options {:init-ns user})


src/number_in_words/conversion.clj
(ns number-in-words.conversion)

(def out-of-boundary-error-message "Error: Out of boundaries - number should be between 0 and 1000 inclusive")

(def one-to-twenty ["zero" "one"  "two"  "three"  "four"  "five"  "six"  "seven"  "eight"  "nine"
                    "ten"  "eleven"  "twelve"  "thirteen"  "fourteen"  "fifteen"  "sixteen"  "seventeen"  "eighteen"  "nineteen"  "twenty"])

(def tens ["zero" "ten" "twenty" "thirty" "forty" "fifty" "sixty" "seventy" "eighty" "ninety"])

(defn- round-tens? [n] (and (< n 100) (= 0 (mod n 10))))
(defn- round-hundreds? [n] (and (< n 1000) (= 0 (mod n 100))))

(defn- extract-units
  "Returns the units for any number n"
  [n]
  (mod n 10))

(defn- extract-tens
  "Returns the tens for the number. Eg 30 for 34, or 90 for 99"
  [n]
  (-> (- n (extract-units n))
      (mod 100)))

(defn- extract-hundreds
  "Returns the hundreds for the number. Eg 100 for 178, or 400 for 444"
  [n]
  (-> (- n (mod n 100))
      (mod 1000)))

(defn convert
  "Converts a number into English words."
  [n]
  (cond (not (<= 0 n 1000)) out-of-boundary-error-message
        (<= 0 n 20) (get one-to-twenty n)
        (round-tens? n) (get tens (/ n 10))
        (<= 21 n 99) (str (convert (extract-tens n)) "-" (convert (extract-units n)))
        (round-hundreds? n) (str (convert (/ n 100)) " hundred")
        (<= 101 n 999) (str (convert (extract-hundreds n)) " and " (convert (mod n 100)))
        (= n 1000) "one thousand"))
